File: Day 10/solution.java
class Solution {
    public int[] findBall(int[][] grid) {
        int[] ans = new int[grid[0].length];
        for(int ball = 0; ball < grid[0].length; ball++)
        {
            int currentPos = ball;
            for(int row = 0; row < grid.length; row++)
            {
                int nextPos = currentPos + grid[row][currentPos];
                if(nextPos < 0 || nextPos >= grid[0].length ||
                  grid[row][currentPos] != grid[row][nextPos])
                {
                    ans[ball] = -1;
                    break;
                }
                ans[ball] = nextPos;
                currentPos = nextPos;
            }
        }
        return ans;
    }
}

File: Day 10/solution.cpp
class Solution {
public:
    vector<int> findBall(vector<vector<int>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        int dp[rows + 1][cols];
        memset(dp, 0, sizeof(dp));
        for(int c = 0; c < cols; c++) dp[rows][c] = c;
        for(int r = rows - 1; r >= 0; r--){
            for(int c = 0; c < cols; c++){
                int nc = c + grid[r][c]; // next column
                if(nc < 0 || nc > cols - 1 || grid[r][c] != grid[r][nc]) dp[r][c] = -1;
                else dp[r][c] = dp[r + 1][nc];
            }
        }
        vector<int> res(cols, 0);
        for(int c = 0; c < cols; c++) res[c] = dp[0][c];
        return res;
    }
};

//1/11/2022